extends Control
class_name Manifestation

@export var types : Dictionary[Enums.ElementType,int]
var god : God
@onready var description :RichTextLabel = $Description/RichTextLabel

func act():
	for type in types :
		if check_elements():
			pay_cost_elements()


func _on_button_mouse_entered() -> void:
	$DescriptionTimer.start(0.7)


func _on_button_mouse_exited() -> void:
	$Description.visible = false
	$DescriptionTimer.stop()


func _on_description_timer_timeout() -> void:
	$Description.visible = true

func check_elements() -> bool:
	for type in types :
		if ! god.check_element(type,types[type]):
			return false
	return true
	
func pay_cost_elements() -> void :
	for type in types :
		god.pay_cost_element(type,types[type])
		
func add_element_self(type : Enums.ElementType, value : int) -> void:
	god.add_element(type, value)

func remove_global_element(type : Enums.ElementType, value : int) -> void:
	god.remove_global_element(type, value)

func multiply_element(type :Enums.ElementType, value : int) -> void :
	god.multiply_element(type,value)
	
func multiply_global_element(type :Enums.ElementType, value : int) -> void :
	god.multiply_global_element(type,value)

func _on_button_button_down() -> void:
	act()
